src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp

```diff
@@ -182,11 +379,107 @@ namespace
             else value.set(0.0f);
         }
 
-        ShaderGroup* m_shader_group;
+        float compute_pdf(
+            const float   u,
+            const float   v,
+            const float   theta) const
+        {
+            assert(u >= 0.0f && u < 1.0f);
+            assert(v >= 0.0f && v < 1.0f);
+            assert(m_importance_sampler);
+
+            // Compute the probability density of this sample in the importance map.
+            size_t m_importance_map_width = m_importance_map_size;
+            size_t m_importance_map_height = m_importance_map_size / 2;
+            const size_t x = truncate<size_t>(m_importance_map_width * u);
+            const size_t y = truncate<size_t>(m_importance_map_height * v);
+            const float prob_xy = m_importance_sampler->get_pdf(x, y);
+            assert(prob_xy >= 0.0f);
+
+            // Compute the probability density of the emission direction.
+            const float pdf = prob_xy > 0.0f ? prob_xy * m_probability_scale / std::sin(theta) : 0.0f;
+            assert(pdf >= 0.0f);
+
+            return pdf;
+        }
+
+        void build_importance_map(const Project& project, IAbortSwitch* abort_switch)
+        {
+            Stopwatch<DefaultWallclockTimer> stopwatch;
+            stopwatch.start();
+
+            size_t m_importance_map_width = m_importance_map_size;
+            size_t m_importance_map_height = m_importance_map_size / 2;
+            m_rcp_importance_map_width = 1.0f / m_importance_map_width;
+            m_rcp_importance_map_height = 1.0f / m_importance_map_height;
+
+            const size_t texel_count = m_importance_map_width * m_importance_map_height;
+            m_probability_scale = texel_count / (2.0f * PiSquare<float>());
+
+            // Create OSL shading system.
+            auto_release_ptr<OIIOTextureSystem> texture_system(
+                OIIOTextureSystemFactory().create());
+
+            RendererServices renderer_services(project, texture_system.ref());
+
```

Comment:
Remove blank line.

The idea is that the comment "Create OSL shading system" really applies to the whole block so it shouldn't be interrupted by blank lines.

Reply:
I understand. Done.